### ISSUE TITLE
onnx to mlir conversion - failed to legalize operation 'onnx.InstanceNormalization'

### ISSUE THREAD

**Issue by kraza8** [onnx/onnx-mlir#3141]
Hi, seeing this issue when converting onnx model to mlir:

 _./onnx-mlir --EmitMLIR /proj/work/kraza/git/diffusion_policy/Model_Zoo/diffusion_policy/noise_pred_net.onnx_
[1/3] Fri Apr 25 14:10:29 2025 (0s) Importing ONNX Model to MLIR Module from "noise_pred_net.onnx"

Warning: ONNX Softplus in your model is using Opset 1, which is quite old. Please consider regenerating your model with a newer Opset.
......
.........
Warning: ONNX Softplus in your model is using Opset 1, which is quite old. Please consider regenerating your model with a newer Opset.

[2/3] Fri Apr 25 14:10:29 2025 (0s) Compiling and Optimizing MLIR Module
failed to legalize operation 'onnx.InstanceNormalization' that was explicitly marked illegal

**

[noise_pred_net_onnx_graph.txt](https://github.com/user-attachments/files/19917132/noise_pred_net_onnx_graph.txt)

**
Linux Ubuntu 22.04*):
torch 2.6.0
Linux Ubuntu 22.04*):
onnx 1.17.0
protobuf 5.29.4

**Code:**
sample = torch.randn(1, 16, 2, dtype=torch.float32)  # (batch_size, seq_len, action_dim)
  timestep = torch.tensor([7], dtype=torch.long)       # diffusion step
  global_cond = torch.randn(1, 1028, dtype=torch.float32)  # (batch_size, obs_horizon * obs_dim)

  onnx_model_path = "noise_pred_net.onnx"  # local path for testing

  try:
    torch.onnx.export(
    ema_nets['noise_pred_net'],
    (sample, timestep, global_cond),
    onnx_model_path,
    input_names=["sample", "timestep", "global_cond"],
    output_names=["predicted_noise"],
    dynamic_axes={
            "sample": {0: "batch_size", 1: "seq_len"},
            "timestep": {0: "batch_size"},
            "global_cond": {0: "batch_size"},
            "predicted_noise": {0: "batch_size", 1: "seq_len"},
        },
        opset_version=17,
        export_params=True,
    )
    print("ONNX export successful! Model saved to:", onnx_model_path)
  except Exception as e:
    print("ONNX export failed:", e)

**Expected Behavior:**
onnx model to generate its MLIR form without any issues.

Onnx model graph attached.

**Comment by AlexandreEichenberger** [comment]
Can you post the actual ".onnx" file? 

Also, the compiler seems to indicate that there are operations from Opset 1, which is very old. I would say that we get good coverage from models encoded from Opset 9 and onwards. However, in this specific case, Opset 1 has been the "newest" release until Opset 22. So it's probably fine. Maybe we should update the minimal supported version for that specific op.
https://onnx.ai/onnx/operators/onnx__Softplus.html

This doc indicates which opset onnx-mlir currently supports (and/or what restrictions we currently have)
https://github.com/onnx/onnx-mlir/blob/main/docs/SupportedONNXOps-cpu.md

**Comment by kraza8** [comment]
Please download the onnx file:
https://tsavoritesi-my.sharepoint.com/:u:/p/kraza/EUUqN3gkYmpGslQqEfRvfMoBMQc41qKbbQrpYXJa5n4u6Q?e=ZHY5uq

**Comment by AlexandreEichenberger** [comment]
Thanks. PR #3148 fixes the issue listed above.

However, the onnx file you have listed has another issue: it list the transpose as below:

```mlir
  %148 = "onnx.Transpose"(%arg0) {onnx_node_name = "/Transpose", perm = [0, -1, 1]} : (tensor<?x?x2xf32>) -> tensor<*xf32>
```

but to the best of my knowledge transpose `perm` accept numbers only between `0` and `rank-1` inclusively. I went and checked what ORT is doing, and they also seem to only accept positive number (feel free to check if you have ORT installed, I don't have a copy handy).

This can be manually fixed, either by editing the .onnx file. In my case, I emitted the .mlir file using `onnx-mlir noise_pred_net.onnx -EmitONNXBasic` and edited the `-1` to their logical equivalent of `2`.

**Comment by AlexandreEichenberger** [comment]
There is still an issue with `ConvTranspose`, and it is the same issue as with the InstanceNorm, namely that we did it in the `decompose` pass but that pass occurs before we have shape inference. So I would need to move these optimizations from decompose to canonicalize. 

Let me do that after the PR above is accepted, so that I don't need to replicate fixes.

**Comment by AlexandreEichenberger** [comment]
I looked into the `ConvTranspose` and here is the rub. The model used spacial dimensions that are dynamic.

```mlir
  %627 = "onnx.ConvTranspose"(%626, %133, %134) {auto_pad = "NOTSET", dilations = [1], group = 1 : si64, kernel_shape = [4], onnx_node_name = "/up_modules.1.2/conv/ConvTranspose", pads = [1, 1], strides = [2]} : (tensor<?x256x?xf32>, tensor<256x256x4xf32>, tensor<256xf32>) -> tensor<?x256x?xf32>
```

At this time, we decompose a `ConvTranspose` into `Conv` plus a couple of operations, one of which is `pad`, which requires the amount of padding to be a literal. With dynamic spacial dimensions, this would be a runtime value.

So the work around is to set the "spacial" dimension (1D here) to a constant value when compiling. This can be done by setting the `onnx-mlir -shapeInformation=xxx` flag.

The general solution would be to implement `ConvTranspose` lowering directly. This requires a significant expertise in convolution (normal one is already tricky, transpose is an additional level of complexity). We would welcome outside contributors on this one, happy to help with the mechanics of how to implement this.

**Comment by kraza8** [comment]
Thanks, 

now running into this issue:

git log --oneline | grep 3148
85ccac70 migrated instance/group normalization from decompose to canonicalize (#3148)

./onnx-mlir --EmitMLIR /proj/work/kraza/git/diffusion_policy/Model_Zoo/diffusion_policy/noise_pred_net.onnx
[1/3] Thu May  8 14:48:53 2025 (0s) Importing ONNX Model to MLIR Module from "noise_pred_net.onnx"

Warning: ONNX Softplus in your model is using Opset 1, which is quite old. Please consider regenerating your model with a newer Opset.

[2/3] Thu May  8 14:48:54 2025 (1s) Compiling and Optimizing MLIR Module
onnx-mlir: /proj/work/kraza/git/onnx-mlir2/onnx-mlir/src/Dialect/Mlir/IndexExprBuilder.cpp:370: int64_t onnx_mlir::IndexExprBuilder::getShape(mlir::Value, uint64_t): Assertion `i < rank && "expected index smaller than memref rank"' failed.

How do you generate .mlir file? I tried with --EmitONNXIR and --EmitMLIR but the above error is preventing it.

**Comment by AlexandreEichenberger** [comment]
If you want to generate an mlir file just after ingesting it from ONNX before any processing, use the `--EmitONNXBasic`. That is strait after the ONNX to MLIR converter before any passes.

**Comment by kraza8** [comment]
Hi, we are still running into this issue:
onnx-mlir: /proj/work/kraza/git/onnx-mlir2/onnx-mlir/src/Dialect/Mlir/IndexExprBuilder.cpp:370: int64_t onnx_mlir::IndexExprBuilder::getShape(mlir::Value, uint64_t): Assertion `i < rank && "expected index smaller than memref rank"' failed.

**Comment by kraza8** [comment]
onnx was generated from pytorch 2.6.0

**Comment by AlexandreEichenberger** [comment]
@kraza8 

we had a re-read of the specs and decided to support the negative numbers (range -rank... rank-1 inclusive).

so you should be able to run the model now without changes. Will need the latest onnx-mlir. Fixed in #3191 

hope this helps.
